### PR TITLE
[NTDLL_APITEST] Expand the NtSetInformationProcess testcase with some tests

### PR DESCRIPTION
**Pull request by Fraizeraust**
`ProcessRaisePriority` class expects a buffer information of **KPRIORITY** type.
**Windows Server 2003**
Capture2
**Windows XP**
Capture4
**Windows 10 Home 1909**
Capture
**ReactOS**
Capture6

### CODE REVIEW

File: modules/rostests/apitests/ntdll/NtSetInformationProcess.c [hunk 1]
```diff
@@ -7,6 +7,162 @@
...
 #include "precomp.h"
 
+static
+void
+Test_ProcBasePriorityClass(void)
+{
+    NTSTATUS Status;
+    KPRIORITY *BasePriority;
+
+    BasePriority = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(KPRIORITY));
```
Comment (HBelusca) [inline modules/rostests/apitests/ntdll/NtSetInformationProcess.c:17]: Why not just using a regular buffer (on the stack)?

File: modules/rostests/apitests/ntdll/NtSetInformationProcess.c [hunk 2]
```diff
@@ -7,6 +7,162 @@
...
+}
+
+static
+void
+Test_ProcRaisePriorityClass(void)
+{
+    NTSTATUS Status;
+    KPRIORITY *RaisePriority;
+
+    RaisePriority = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(KPRIORITY));
```
Comment (HBelusca) [inline modules/rostests/apitests/ntdll/NtSetInformationProcess.c:103]: Ditto.

Comment (HBelusca) [reply modules/rostests/apitests/ntdll/NtSetInformationProcess.c:103]: Isn't it also a ULONG here?

Comment (Fraizeraust) [reply modules/rostests/apitests/ntdll/NtSetInformationProcess.c:103]: No? **ProcessRaisePriority** wants a buffer of KPRIORITY which is a LONG but at the same time it wants a length information of ULONG.

Comment (HBelusca) [reply modules/rostests/apitests/ntdll/NtSetInformationProcess.c:103]: Then it's simply a ULONG that it wants 😄

**Comment by HBelusca** [comment]
For notice, our kernel code:
- for ProcessBasePriority:
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ps/query.c;hb=1f3cb13fb97e618b53e55db0f455fe7259694326#l1525

- for ProcessRaisePriority:
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ps/query.c;hb=1f3cb13fb97e618b53e55db0f455fe7259694326#l1589

appears to do what's necessary already.

**Comment by HBelusca** [comment]
The error codes that the test running in ReactOS does mention:
```
/error 0xc0000004:
    STATUS_INFO_LENGTH_MISMATCH

/error 0xc0000008:
    STATUS_INVALID_HANDLE

/error 0x80000002:
    STATUS_DATATYPE_MISALIGNMENT
```

**Comment by HBelusca** [comment]
Basically, what the test run shows on ReactOS, is that the `NtSetInformationProcess()` function misses some ProbeForRead calls, similar to what is done here:
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ps/query.c;hb=1f3cb13fb97e618b53e55db0f455fe7259694326#l88
(this was originally what was intended by the commented-out call to `DefaultSetInfoBufferCheck()`).

Also I've checked, and it appears as well that `NtSetInformationThread()` misses these probings too

**Comment by Fraizeraust** [comment]
bf493b9 is the commit where `ps_i.h` tables were removed and c4f9542 where the probing checks were disabled.

File: modules/rostests/apitests/ntdll/NtSetInformationProcess.c [hunk 3]
```diff
@@ -7,6 +7,140 @@
...
+    Status = NtSetInformationProcess(NtCurrentProcess(),
+                                     ProcessBasePriority,
+                                     (PVOID)1,
+                                     0);
+    ok_hex(Status, STATUS_INFO_LENGTH_MISMATCH);
+
+    /* The input buffer is misaligned */
+    Status = NtSetInformationProcess(NtCurrentProcess(),
+                                     ProcessBasePriority,
+                                     (PVOID)1,
```
Comment (ThFabba) [inline modules/rostests/apitests/ntdll/NtSetInformationProcess.c:62]: For completeness, test `2` as well to show that the alignment is indeed four and not two (which is unlikely, but still...)

File: modules/rostests/apitests/ntdll/NtSetInformationProcess.c [hunk 4]
```diff
@@ -7,6 +7,140 @@
...
+                                     ProcessBasePriority,
+                                     &BasePriority,
+                                     sizeof(KPRIORITY));
+    ok_hex(Status, STATUS_PRIVILEGE_NOT_HELD);
+
+    /*
+     * Assign a random priority value this time and
+     * set the base priority to the current process.
+    */
+    BasePriority = 8;
```
Comment (ThFabba) [inline modules/rostests/apitests/ntdll/NtSetInformationProcess.c:77]: For the future: the interesting test cases are boundary values -- i.e. what is the last value that succeeds and what is the first that returns `STATUS_PRIVILEGE_NOT_HELD`?

File: modules/rostests/apitests/ntdll/NtSetInformationProcess.c [hunk 5]
```diff
@@ -7,6 +7,140 @@
...
+    Status = NtSetInformationProcess(NtCurrentProcess(),
+                                     ProcessBasePriority,
+                                     &BasePriority,
+                                     sizeof(KPRIORITY));
+    ok_hex(Status, STATUS_SUCCESS);
+}
```
Comment (ThFabba) [inline modules/rostests/apitests/ntdll/NtSetInformationProcess.c:83]: Because these tests run one after the other in the same process, each test should leave the process in the state it found it in (i.e. this function should query the priority at the beginning, then reset it back at the end). Otherwise test results might start to depend on things done by previous tests, and changing the order would change results.

File: modules/rostests/apitests/ntdll/NtSetInformationProcess.c [hunk 6]
```diff
@@ -7,6 +7,147 @@
...
+    Status = NtSetInformationProcess(NtCurrentProcess(),
+                                     ProcessRaisePriority,
+                                     NULL,
+                                     sizeof(ULONG));
+    ok_hex(Status, STATUS_ACCESS_VIOLATION);
+
+    /* The input buffer is misaligned */
+    Status = NtSetInformationProcess(NtCurrentProcess(),
+                                     ProcessRaisePriority,
+                                     (PVOID)1,
```
Comment (ThFabba) [inline modules/rostests/apitests/ntdll/NtSetInformationProcess.c:139]: Add `(PVOID)2` test?

File: modules/rostests/apitests/ntdll/NtSetInformationProcess.c [hunk 7]
```diff
@@ -7,6 +7,147 @@
...
+
+    /* The input buffer is misaligned and length information invalid */
+    Status = NtSetInformationProcess(NtCurrentProcess(),
+                                     ProcessRaisePriority,
+                                     (PVOID)1,
+                                     0);
+    ok_hex(Status, STATUS_INFO_LENGTH_MISMATCH);
+
+    /* The information length is invalid */
+    Status = NtSetInformationProcess(NtCurrentProcess(),
```
Comment (ThFabba) [inline modules/rostests/apitests/ntdll/NtSetInformationProcess.c:123]: This doesn't seem to add anything over the one just above.
